File: morse2/src/morse.zig
const std = @import("std");

const ArrayList = std.ArrayList;
const allocator = std.testing.allocator;

const map = @import("mapping.zig");

pub fn encode(message: []const u8) ![]const u8 {
    var mapping = try map.morseAlphaMap(allocator);
    defer mapping.deinit();
    var encoded = ArrayList(u8).init(allocator);

    for (message) |char| {
        if (mapping.get(char)) |code| {
            for (code) |codeChar| {
                try encoded.append(codeChar);
            }
            try encoded.append(' ');
        }
    }

    _ = encoded.pop();

    return encoded.toOwnedSlice();
}

pub fn decode(code: []const u8) ![]u8 {
    var mapping = try map.alphaMorseMap(allocator);
    defer mapping.deinit();
    var chars = ArrayList(u8).init(allocator);
    var iter = std.mem.split(u8, code, " ");

    while (iter.next()) |c| {
        if (mapping.get(c)) |char| {
            try chars.append(char);
        }
    }

    return chars.toOwnedSlice();
}

test "encode" {
    var e = try encode("hello world!");
    defer allocator.free(e);

    try std.testing.expectEqualStrings(e, ".... . .-.. .-.. --- / .-- --- .-. .-.. -.. -.-.--");
}

test "decode" {
    var d = try decode(".... . .-.. .-.. --- / .-- --- .-. .-.. -.. -.-.--");
    defer allocator.free(d);

    try std.testing.expectEqualStrings(d, "hello world!");
}

File: morse-rs/src/morse.rs
use crate::mapping;

pub fn encode(message: &str) -> String {
    let mapping = mapping::morse_alpha_map();
    let mut encoded = message.chars().fold(String::new(), |mut s, c| {
        let d = match mapping.get(&c) {
            None => "",
            Some(ch) => ch,
        };

        s.push_str(d);
        s.push_str(" ");
        s
    });

    // remove last space
    encoded.pop();

    encoded
}

pub fn decode(code: &str) -> String {
    let mapping = mapping::alpha_morse_map();
    code.split(' ').fold(String::new(), |mut s, symbol| {
        match mapping.get(symbol) {
            None => (),
            Some(ch) => s.push(*ch),
        };

        s
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_works() {
        let m = "hello world!";
        let e = encode(m);
        assert_eq!(e, ".... . .-.. .-.. --- / .-- --- .-. .-.. -.. -.-.--");
    }

    #[test]
    fn decode_works() {
        let c = ".... . .-.. .-.. --- / .-- --- .-. .-.. -.. -.-.--";
        let d = decode(c);
        assert_eq!(d, "hello world!");
    }
}
